File: algorithms/2_medium/M/03002_maximum_size_of_a_set_after_removals/hashset.rs
use std::collections::HashSet;

impl Solution {
    pub fn maximum_set_size(nums1: Vec<i32>, nums2: Vec<i32>) -> i32 {
        assert!(nums1.len() == nums2.len());
        let len = nums1.len();
        assert!(len > 0 && (len & 1) == 0);

        let half = len / 2;
        let set1: HashSet<i32> = nums1.into_iter().collect();
        let set2: HashSet<i32> = nums2.into_iter().collect();

        let mut cnt1 = 0_usize;
        for num in &set1 {
            if !set2.contains(num) {
                cnt1 += 1;
                if cnt1 == half {
                    break;
                }
            }
        }

        let mut cnt2 = 0_usize;
        for num in &set2 {
            if !set1.contains(num) {
                cnt2 += 1;
                if cnt2 == half {
                    break;
                }
            }
        }

        let mut cnt = cnt1 + cnt2;
        if cnt < len {
            for num in &set1 {
                if set2.contains(num) {
                    cnt += 1;
                    if cnt == len {
                        break;
                    }
                }
            } 
        }

        cnt.try_into().unwrap()
    }
}

File: algorithms/1_easy/S/03718_smallest_missing_multiple_of_k/hashset.cpp
class Solution {
public:
    static int missingMultiple(const vector<int>& nums, const int k) {
        assert(k > 0);
        auto hashset = unordered_set<int>();

        for (const int num : nums) {
            if (num % k == 0) {
                hashset.insert(num);
            }
        }

        int curr = k;

        while (hashset.find(curr) != hashset.end()) {
            curr += k;
        }

        return curr;
    }
};
